### DISCUSSION TITLE
Feedback: Have you used the `aarch64-linux` native gem?

### DISCUSSION THREAD

**Discussion by flavorjones** [sparklemotion/nokogiri#2359] [Releases]
An experimental version of Nokogiri precompiled for ARM64 Linux platforms is available at:

https://storage.googleapis.com/rubygem-nokogiri-alphas/nokogiri-1.13.0.alpha2-aarch64-linux.gem

(see details at https://github.com/sparklemotion/nokogiri/pull/2332), and is planned to be available in the next point release of Nokogiri.

If you've tried the native gem on ARM64 Linux, please comment here! We don't have CI automatically testing on this platform yet, so my confidence is low pending feedback from y'all that this is working for you.

**Comment by navied** [comment]
Just gave the new release a try, I am running into this bug with this on AWS Codebuild ARM and M1 Macs:

```
LoadError: Error loading shared library ld-linux-aarch64.so.1: No such file or directory (needed by /codebuild/output/src237909456/src/vendor/bundle/ruby/2.7.0/gems/nokogiri-1.13.0-aarch64-linux/lib/nokogiri/2.7/nokogiri.so) - /codebuild/output/src237909456/src/vendor/bundle/ruby/2.7.0/gems/nokogiri-1.13.0-aarch64-linux/lib/nokogiri/2.7/nokogiri.so
```

**Comment by flavorjones** [comment]
Hi @navied, thanks for the feedback.

The M1 Mac is am arm64-darwin platform, and so I don't expect the aarch64-linux gem to work there. Pardon me if I've misunderstood what you're reporting.

I'm not familiar with AWS Codebuild ARM, so can I ask you to [open a new issue](https://github.com/sparklemotion/nokogiri/issues/new?assignees=&labels=topic%2Finstallation&template=1-installation-difficulties.md&title=%5Binstall%5D) and provide me with enough information to reproduce what you're seeing? Thanks very much!

**Reply by navied** [reply]
Hi @flavorjones thanks for the quick response, AWS Codebuild uses Graviton underneath (which the changelog mentions it should support). As for the M1 mac, I should have mentioned that the problem I am seeing is through docker which utilized arm64 images underneath.

But I actually believe I have honed in the on the issue, it seems to be a problem on alpine based images, if I had to guess we need musl compiled packages for it to work on alpine.

I can make an issue outlining how to reproduce if you'd like! But it as simple as installing nokogiri on an alpine based image on arm and it will fail to load.

Edit: created issue here: https://github.com/sparklemotion/nokogiri/issues/2414

**Reply by flavorjones** [reply]
#2414 caused us to update Nokogiri's installation documentation to include a [musl-specific section](https://nokogiri.org/tutorials/installing_nokogiri.html#linux-musl-error-loading-shared-library) that instructs users to `apk add gcompat`. This may not have been needed previously by aarch64-linux users, but with Nokogiri v1.13.0 shipping native (precompiled) gems for that platform, it now is.

**Reply by Nakilon** [reply]
Adding `gcompat` does not help me. I'm switching from Ubuntu VM to M4.
```
LoadError:
  cannot load such file -- nokogiri/nokogiri
...
```

**UPD:** nevermind, I didn't notice I had two docker-compose files to edit. This indeed fixes the issue.

**Reply by flavorjones** [reply]
@Nakilon Please note that as of Nokogiri v1.18.0, there are separate native gems for GNU and Musl, meaning that gcompat is not needed for Nokogiri if you're running on a v1.18 release.

**Reply by Nakilon** [reply]
> @Nakilon Please note that as of Nokogiri v1.18.0, there are separate native gems for GNU and Musl, meaning that gcompat is not needed for Nokogiri if you're running on a v1.18 release.

Looks like it does not resolve above 1.15 on ruby 2.7

**Reply by flavorjones** [reply]
> Looks like it does not resolve above 1.15 on ruby 2.7

Correct, support for Ruby 2.7 was dropped as of v1.16.0 in December 2023.

**Comment by ajpalok** [comment]
I'm using aarch-linux where this issue have been created with `Nokogiri 1.13.0` #2416

**Comment by flavorjones** [comment]
Just an update that https://github.com/sparklemotion/nokogiri/pull/2425 introduced test coverage for aarch64-linux, so we're green. Thanks for the feedback, everybody.

**Comment by joeyparis** [comment]
I am having trouble starting `rails 7` with `Nokogiri 1.13.0` on `Ruby 3.0 running on 64bit Amazon Linux 2/3.4.1`. I get this error on startup:

```
ERROR: It looks like you're trying to use Nokogiri as a precompiled native gem on a system with glibc < 2.17:

  /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /var/app/current/vendor/bundle/ruby/3.0.0/gems/nokogiri-1.13.0-aarch64-linux/lib/nokogiri/3.0/nokogiri.so) - /var/app/current/vendor/bundle/ruby/3.0.0/gems/nokogiri-1.13.0-aarch64-linux/lib/nokogiri/3.0/nokogiri.so

  If that's the case, then please install Nokogiri via the `ruby` platform gem:
      gem install nokogiri --platform=ruby
  or:
      bundle config set force_ruby_platform true

  Please visit https://nokogiri.org/tutorials/installing_nokogiri.html for more help.
```

The error message isn't 100% accurate either. I'm not attempting to use Nokogiri with glibc < 2.17, I'm trying to run it with `GLIBC_2.26`, but the error message is triggered by. Nokogiri is not finding 2.29 as only 2.26 is installed (still above 2.17).

```
ruby:
      version: 3.0.3
      platform: aarch64-linux
      gem_platform: aarch64-linux
      description: ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [aarch64-linux]
      engine: ruby
```

**Reply by joeyparis** [reply]
Okay, upon further investigation it looks like the [minimum GLIBC version for ARM Linux platforms is actually 2.29](https://github.com/sparklemotion/nokogiri/blob/be178293898e1f0002df0477fc7d33a6841aac2b/rakelib/extensions.rake#L193). So really [this error message](https://github.com/sparklemotion/nokogiri/blob/913efce3087b40f0c700b95cc503bc232fb47fed/lib/nokogiri/extension.rb) just needs to be updated to reflect the correct minimum version on ARM architectures.

On that note (I haven't investigated further yet), is it at all possible to make the minimum version be 2.26 as that's the current default on Amazon Linux 2 managed platforms right now?

**Reply by flavorjones** [reply]
@joeyparis Thanks for commenting. I've created a new issue to clarify the error message emitted by  `lib/nokogiri/extension.rb` at https://github.com/sparklemotion/nokogiri/issues/2427

I'll double-check that v2.29 is really the minimum on arm64-linux, but please note that this is a requirement put on the precompiled native gem by `rake-compiler-dock` and there may not be anything we can easily do from within Nokogiri to change this. CC @larskanis

**Reply by flavorjones** [reply]
@joeyparis Confirmed, `aarch64-linux` requires glibc 2.29 or higher to ensure everything will work.

Here's how we can see this: download the precompiled gems for x86_64-linux and aarch64-lnux:

```sh
gem fetch nokogiri --platform=x86_64-linux
gem fetch nokogiri --platform=aarch64-linux
```

Open up the archives (or install them if that's easier) and take a look at `lib/nokogiri/3.0/nokogiri.so` for each using `objdump -T`:

```text
$ objdump -T nokogiri.so-aarch64-linux | fgrep GLIBC | cut -c50-60 | sort | uniq -c
    108 GLIBC_2.17 
      1 GLIBC_2.28 
      3 GLIBC_2.29 

$ objdump -T nokogiri.so-x86_64-linux | fgrep GLIBC | cut -c50-60 | sort | uniq -c
      1 GLIBC_2.14 
      1 GLIBC_2.17 
    102 GLIBC_2.2.5
      4 GLIBC_2.3  
      4 GLIBC_2.3.2
      1 GLIBC_2.4  
      1 GLIBC_2.7  
```

We really need to pin the required version to the highest symbol version we see.

**Reply by flavorjones** [reply]
@joeyparis In the meantime, you should follow the instructions at https://nokogiri.org/tutorials/installing_nokogiri.html#how-can-i-avoid-using-a-precompiled-native-gem to compile Nokogiri from source, it's only the precompiled native gem that isn't supported with older versions of glibc.

**Reply by joeyparis** [reply]
@flavorjones Thanks for the help! I was able to successfully do this for the `active-elastic-job` gem by adding this [prebuild hook](https://github.com/active-elastic-job/active-elastic-job/pull/138/files#diff-261ceb8cf3c96d974873c3b3667206aded79b297774ccebdc76467ddb55d308f). I believe this will work for anyone attempting to install Nokogiri on an Amazon Linux 2 Elastic Beanstalk platform until they default to using glibc 2.29.

**Reply by rkiller** [reply]
Just in case someone else stumbles on @joeyparis link and solution... in order for that to work you must 'bundle lock --add-platform ruby' to enable the ruby platform in your Gemfile.lock

**Reply by joeyparis** [reply]
@rkiller you are correct! I had already run `bundle lock --add-platform ruby` early on in debugging this so I didn't remember that being relevant. Thanks for the catch!

**Reply by januszm** [reply]
Hi @joeyparis , did you find a solution how to run Rails 7 on AL2 t4g instance `aarch64-linux` ?

**Reply by joeyparis** [reply]
@januszm yes, I was able to get it working with this [prebuild hook](https://github.com/active-elastic-job/active-elastic-job/pull/138/files#diff-261ceb8cf3c96d974873c3b3667206aded79b297774ccebdc76467ddb55d308f). Coincidentally, on a Ruby gem we're both maintainers of haha

**Reply by januszm** [reply]
Ah that one! Yes, great, thank you 😉 .  Newer GLIBC will only be included in Amazon Linux 2022+ it seems, so I guess there's no hope to have it on AL2 without workarounds.

**Reply by joeyparis** [reply]
Unfortunately, that seems to be the case with a handful of aarch64/Ruby 3 benefits. I've been able to work around most of them, but have a lot of prebuild hooks for them. Thanks for the info about waiting for the Amazon Linux 2022+, I didn't know about that!

**Reply by brcebn** [reply]
Hi @joeyparis 
I got a similar issue with Ruby `2.7.2` and Rails `6.1.6`. I'm using rvm and it seems bundle config does not apply when running a Rake task. 

My task contains
```ruby
Webpacker.with_node_env('production') do
  if MyEngine.webpacker.commands.compile
    # Successful compilation!
  else
    # Failed compilation
    exit!
  end
end
```

```ruby
# my_engine/lib/my_engine.rb
# frozen_string_literal: true

require 'my_engine/engine'

module MyEngine # rubocop:disable Style/Documentation
  class << self
    def webpacker
      @webpacker ||= ::Webpacker::Instance.new(
        root_path: MyEngine::Engine.root,
        config_path: MyEngine::Engine.root.join('config/webpacker.yml')
      )
    end
  end
end
```

Do you have any idea to fix this? 🙂

**Reply by brcebn** [reply]
Obviously, that's when you write down the message that you find the solution from the [documentation](https://nokogiri.org/tutorials/installing_nokogiri.html#installing-using-the-packaged-libraries)

```
bundle config force_ruby_platform true
```

**Reply by flavorjones** [reply]
If you're using a recent Bundler you can do this per-gem, so your Gemfile would have:

```
# Gemfile
gem "nokogiri", force_ruby_platform: true
```

**Reply by brcebn** [reply]
I'm current using `2.0.1` but upgrade is coming. Thanks for the tip! 👌🏻

**Reply by Ahmed-Salama96** [reply]
same here with `Ruby 2.7 running on 64bit Amazon Linux 2/3.6.1` i have another app on ruby 2.7.5 and working fine but now aws runs on 2.7.7 so I cannot start a new project because of this error
<img width="1620" alt="Screenshot 2022-12-25 at 12 19 11 PM" src="https://user-images.githubusercontent.com/29269770/209456838-efb82b29-d1db-47a9-b7e8-f0e758796451.png">

**Reply by flavorjones** [reply]
@Ahmed-Salama96 please read the thread for workarounds.

**Reply by theRocket** [reply]
the v14.0 release notes say:

> Faster, more reliable installation: Native Gem for aarch64-linux (aka linux/arm64/v8)
This version of Nokogiri ships official native gem support for the aarch64-linux platform, which should support AWS Graviton and other ARM64 Linux platforms. Please note that glibc >= 2.29 is required for aarch64-linux systems, see [Supported Platforms](https://nokogiri.org/#supported-platforms) for more information.

I shipped this version to an AWS Lambda layer and I'm still getting this error:
> /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /opt/ruby/2.7.0/gems/nokogiri-1.14.1-aarch64-linux/lib/nokogiri/2.7/nokogiri.so) - /opt/ruby/2.7.0/gems/nokogiri-1.14.1-aarch64-linux/lib/nokogiri/2.7/nokogiri.so

I am using Lambda arm64 compatible images to build the gem and selecting arm64 processing for my lambda and the layer. So what am I missing?
<img width="1242" alt="AWSLamba_RJRBaseballDelivery_ArmArch" src="https://user-images.githubusercontent.com/5635/216707842-963807cd-2260-4a1c-a72d-0d354f0b5596.png">

**Reply by theRocket** [reply]
I think I found my answer here:
https://repost.aws/questions/QUrXOioL46RcCnFGyELJWKLw/glibc-2-27-on-amazon-linux-2
Lambdas are still running Amazon Linux 2
![AmazonLinux2022_glibc2 35](https://user-images.githubusercontent.com/5635/216712273-311bec08-c0d6-4bd4-b5ea-fc0d48880fc2.png)

**Reply by flavorjones** [reply]
@theRocket Thanks for the note, sorry you're having problems, but yes, you're correct, it looks like Amazon Linux 2 does not have a new-enough version of glibc to use the precompiled binary version.
